### DISCUSSION TITLE
Is it possible to share prisma client as a private npm package AND use it to set up a database in docker?

### DISCUSSION THREAD

**Discussion by secondubly** [prisma/prisma#23335] [Q&A]
### Question

I know that this [question exists here](https://github.com/prisma/prisma/discussions/8936) but unfortunately it does not answer what I am trying to do. I have two projects: A and B that both use a private npm package that contains my schema.prisma and prisma client. Locally this is not a problem because I can just run prisma commands whenever.

My problem arises when I try to set these projects up via Docker - I do not know how to get `prisma generate` working in tandem with the docker containers as I have them currently. Is this even possible? Since docker builds don't have access to the completely separate postgres container at build time, any attempts to access them via the database url fails as far as I've tried. If this explanation doesn't make sense, I wrote out this crude diagram to explain it: 

```
Project A (Node project, setup as a docker image)
     - contains `project-db` as a dependency in package.json

Project B (Separate node project, setup as a docker image)
     - contains `project-db` as a dependency in package.json

postgres (setup as a docker service)
     - what prisma uses as the database

project db (private repository published to the npm registry)
     - contains prisma schema
     - has install scripts for prisma migrate and prisma generate
```
I feel like this should be possible but I am not well versed enough in Docker or prisma to get it done. Could anyone provide some insight? I can provide code samples of the various configs I have if they are needed.

**Comment by ludralph** [comment]
Hi @secondubly 👋 

Thank you for rasing this question.

Can you try using a multi-stage build with volume mounting in docker?

The first stage installs dependencies and generates the Prisma client within the container itself. Then, a second stage copies the generated client code and your application code into a smaller image for production. During the first stage, you can mount a volume containing your private project-db package and the prisma CLI.
```yaml
# Stage 1: Build dependencies (including project-db) and generate Prisma client
FROM node:16-alpine AS builder
WORKDIR /app

COPY package*.json ./
RUN npm install

# Mount your private package and prisma CLI
VOLUME /project-db
VOLUME /usr/local/bin/prisma

COPY project-db /project-db
RUN prisma generate

# Copy generated client and your application code
COPY . .

# Stage 2: Production image
FROM node:16-alpine
WORKDIR /app

COPY --from=builder . .

CMD ["npm", "start"]
```

The `builder` stage installs dependencies, including your private `project-db` package mounted from a volume. It also mounts the Prisma CLI, potentially from another volume or by installing it within the container. The `prisma generate` command is executed within the `builder` stage. The final stage copies the generated client code and your application code from the builder stage into the production image.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
